### PR TITLE
Fixes the energy issues, implements the color detection and introduces notification

### PR DESCRIPTION
**Pull request by restelrich**

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -1,3 +1,3 @@
 [submodule "nrf51-sdk"]
 	path = nrf51-sdk
-	url = git@github.com:relayr/nrf51-sdk.git
+	url = git@github.com:restelrich/nrf51-sdk.git
```
Comment (tralamazza) [inline .gitmodules]: nope

File: wunderbar/motion/motion.c [hunk 2]
```diff
@@ -53,18 +96,36 @@ motion_init(struct motion_ctx *ctx)
...
         simble_srv_register(ctx);
 }
 
-
-static struct motion_ctx motion_ctx;
+static void
+notif_timer_cb(struct rtc_ctx *ctx)
+{
+        void *val = &motion_ctx.motion_value;
+	uint16_t len = sizeof(&motion_ctx.motion_value);
```
Comment (tralamazza) [inline wunderbar/motion/motion.c]: indent

File: wunderbar/noiselvl/noiselvl.c [hunk 3]
```diff
@@ -68,31 +86,23 @@ adc_read_blocking()
...
-	nrf_gpio_pin_write(noise_level_pin_OPAMP, value);
-	nrf_gpio_pin_write(noise_level_pin_SWITCH_ON, value);
-}
-
-static void
 noiselvl_connected(struct service_desc *s)
 {
 	enable_converter(true);
-	// TODO add 50ms delay
+	nrf_delay_us(75000);
```
Comment (tralamazza) [inline wunderbar/noiselvl/noiselvl.c]: can't we use a timer instead?

File: wunderbar/noiselvl/noiselvl.c [hunk 4]
```diff
@@ -100,32 +110,78 @@ static void
...
 	enable_converter(false);
+	rtc_update_cfg(noiselvl_ctx.sampling_period, (uint8_t)NOTIF_TIMER_ID, false);
 }
 
 static void
 noiselvl_read_cb(struct service_desc *s, struct char_desc *c, void **val, uint16_t *len)
 {
 	struct noiselvl_ctx *ctx = (struct noiselvl_ctx *) s;
+	enable_converter(true);
+	nrf_delay_us(75000);
```
Comment (tralamazza) [inline wunderbar/noiselvl/noiselvl.c]: can't we use a timer instead?

File: wunderbar/proximity/proximity.c [hunk 5]
```diff
@@ -40,9 +56,55 @@ proximity_read(struct service_desc *s, struct char_desc *c, void **valp, uint16_
 {
         struct proximity_ctx *ctx = (void *)s;
 
+        enable_i2c();
+        tcs3771_init();
+        while(nrf_gpio_pin_read(TCS37717_INT_PIN) == 1);
```
Comment (tralamazza) [inline wunderbar/proximity/proximity.c]: replace by:
```c
while (nrf_gpio_pin_read(TCS37717_INT_PIN) == 1) {
        /* NOTHING */
}
```

File: wunderbar/proximity/proximity.c [hunk 6]
```diff
@@ -40,9 +56,55 @@ proximity_read(struct service_desc *s, struct char_desc *c, void **valp, uint16_
...
+        if (status & BLE_GATT_HVX_NOTIFICATION)
+                rtc_update_cfg(ctx->sampling_period, (uint8_t)NOTIF_TIMER_ID, true);
+        else     //disable NOTIFICATION_TIMER
+                rtc_update_cfg(ctx->sampling_period, (uint8_t)NOTIF_TIMER_ID, false);
+}
+
+static void
+proximity_notif_timer_cb(struct rtc_ctx *ctx)
+{
+        void *val = &proximity_ctx.proximity_value;
```
Comment (tralamazza) [inline wunderbar/proximity/proximity.c]: indent

File: wunderbar/proximity/proximity.c [hunk 7]
```diff
@@ -53,31 +115,159 @@ proximity_init(struct proximity_ctx *ctx)
...
+        // Resolution: 1ms, max value: 16777216 (4 hours)
+        // A value of 0 will disable periodic notifications
+        simble_srv_char_attach_format(&ctx->sampling_period_char,
+		BLE_GATT_CPF_FORMAT_UINT24, 0, ORG_BLUETOOTH_UNIT_UNITLESS);
         ctx->disconnect_cb = proximity_disconnected;
+        ctx->proximity.read_cb = proximity_read;
+        ctx->proximity.notify = 1;
+        ctx->proximity.notify_status_cb = proximity_notify_status_cb;
+        ctx->sampling_period_char.read_cb = proximity_sampling_period_read_cb;
+	ctx->sampling_period_char.write_cb = proximity_sampling_period_write_cb;
```
Comment (tralamazza) [inline wunderbar/proximity/proximity.c]: use some auto indent tool

File: wunderbar/proximity/proximity.c [hunk 8]
```diff
@@ -53,31 +115,159 @@ proximity_init(struct proximity_ctx *ctx)
...
+static void
+rgb_update(struct rgb_ctx *ctx, uint8_t *val)
+{
+        simble_srv_char_update(&ctx->rgb, val);
+}
+
+static void
+rgb_disconnected(struct service_desc *s)
+{
+        rtc_update_cfg(rgb_ctx.sampling_period, (uint8_t)NOTIF_TIMER_ID+1, false);
```
Comment (tralamazza) [inline wunderbar/proximity/proximity.c]: what's the +1?

File: wunderbar/proximity/proximity.c [hunk 9]
```diff
@@ -53,31 +115,159 @@ proximity_init(struct proximity_ctx *ctx)
...
+
+static void
+rgb_read(struct service_desc *s, struct char_desc *c, void **valp, uint16_t *lenp)
+{
+        struct rgb_ctx *ctx = (void *)s;
+
+        enable_i2c();
+        tcs3771_init();
+        nrf_gpio_pin_write(WLED_CTRL_PIN, true);
+        while(nrf_gpio_pin_read(TCS37717_INT_PIN) == 1);
```
Comment (tralamazza) [inline wunderbar/proximity/proximity.c]: use the `/* NOTHING */` in a block

File: wunderbar/proximity/proximity.c [hunk 10]
```diff
@@ -53,31 +115,159 @@ proximity_init(struct proximity_ctx *ctx)
...
+static void
+rgb_sampling_period_write_cb(struct service_desc *s, struct char_desc *c,
+        const void *val, const uint16_t len)
+{
+	struct rgb_ctx *ctx = (struct rgb_ctx *)s;
+        if (*(uint32_t*)val > MIN_SAMPLING_PERIOD)
+                ctx->sampling_period = *(uint32_t*)val;
+        else
+                ctx->sampling_period = MIN_SAMPLING_PERIOD;
+        rtc_update_cfg(ctx->sampling_period, (uint8_t)NOTIF_TIMER_ID+1, true);
```
Comment (tralamazza) [inline wunderbar/proximity/proximity.c]: replace all these +1, +3 with proper constants

File: wunderbar/proximity/proximity.c [hunk 11]
```diff
@@ -53,31 +115,159 @@ proximity_init(struct proximity_ctx *ctx)
...
+
+static void
+rgb_init(struct rgb_ctx *ctx)
+{
+        simble_srv_init(ctx, simble_get_vendor_uuid_class(), VENDOR_UUID_SENSOR_SERVICE_2);
+        simble_srv_char_add(ctx, &ctx->rgb,
+                     simble_get_vendor_uuid_class(), VENDOR_UUID_COLOR_CHAR,
+                     u8"RGB",
+                     8);
+        /* srv_char_attach_format(&ctx->proximity, */
```
Comment (tralamazza) [inline wunderbar/proximity/proximity.c]: remove code comments

File: wunderbar/temp_rh/htu21.c [hunk 12]
```diff
@@ -16,10 +18,16 @@ static inline uint8_t rotr(uint8_t value, uint8_t shift)
...
 {
 	uint8_t result[3];
-	if (!(twi_master_transfer(HTU21_ADDRESS, &cmd, 1, TWI_DONT_ISSUE_STOP) &&
-		twi_master_transfer(HTU21_ADDRESS | TWI_READ_BIT, result, 3, TWI_ISSUE_STOP))) {
+	enable_i2c();
+	nrf_delay_us(15000);
+	if (!(twi_master_transfer(HTU21_ADDRESS, &cmd, 1, TWI_DONT_ISSUE_STOP))) {
 		return false;
 	}
+	nrf_delay_us(50000);
```
Comment (tralamazza) [inline wunderbar/temp_rh/htu21.c]: 50ms?!

File: wunderbar/temp_rh/temp_rh.c [hunk 13]
```diff
@@ -114,22 +219,31 @@ temp_init(struct temp_ctx *ctx)
...
 static void
-notif_timer_cb(struct rtc_ctx *ctx)
+temp_notif_timer_cb(struct rtc_ctx *ctx)
 {
-  simble_srv_char_notify(&temp_ctx.temp, false, 1, &temp_ctx.last_reading);
-	simble_srv_char_notify(&rh_ctx.temp, false, 1, &rh_ctx.last_reading);
+	void *val = &temp_ctx.last_reading;
+	uint16_t len = sizeof(&temp_ctx.last_reading);
+	temp_read_cb(&temp_ctx, &temp_ctx.temp, &val, &len);
+	simble_srv_char_notify(&temp_ctx.temp, false, 1, &temp_ctx.last_reading);
```
Comment (tralamazza) [inline wunderbar/temp_rh/temp_rh.c]: you have `len`, better use it

File: wunderbar/motion/motion.c (hunk 2 above)
Comment (restelrich) [inline wunderbar/motion/motion.c]: same thing than in the sdk. i'll replace the hard tabs for soft tabs.

File: .gitmodules (hunk 1 above)
Comment (restelrich) [inline .gitmodules]: sure nope.

File: wunderbar/noiselvl/noiselvl.c (hunk 3 above)
Comment (restelrich) [inline wunderbar/noiselvl/noiselvl.c]: i'll change the delay for CONV_WAKEUP_TIME

File: wunderbar/proximity/proximity.c (hunk 8 above)
Comment (restelrich) [inline wunderbar/proximity/proximity.c]: in some sensors there are 2 different services, hence, 2 different notifications. i'll change it to NOTIF_TIMER_ID_PROX and NOTIF_TIMER_ID_RGB. same for the temp/rh.

File: wunderbar/temp_rh/htu21.c (hunk 12 above)
Comment (restelrich) [inline wunderbar/temp_rh/htu21.c]: http://www.meas-spec.com/downloads/HTU21D.pdf

page 22, Table "TEMPERATURE"
max. measuring time for temp. sensor with 14 bit resolution

File: wunderbar/temp_rh/temp_rh.c (hunk 13 above)
Comment (tralamazza) [inline wunderbar/temp_rh/temp_rh.c]: actually size of a pointer is probably not useful

**Pull request merged**